#  Indexing, Slicing, Subsetting and Append DataFrames in Python

## Questions

How can I access specific data within my data set?

How can Python and Pandas help me to analyse my data?

## Objectives

Describe what 0-based indexing is.

Manipulate and extract data using column headings and index locations.

Employ slicing to select sets of data from a DataFrame.

Employ label and integer-based indexing to select ranges of data in a dataframe.

Reassign values within subsets of a DataFrame.

Create a copy of a DataFrame.

Query / select a subset of data using a set of criteria using the following operators: =, !=, >, <, >=, <=.

Locate subsets of data using masks.

Describe BOOLEAN objects in Python and manipulate data using BOOLEANs.

We often want to work with subsets of a DataFrame object. There are different ways to accomplish this including: using labels (column headings), numeric ranges, or specific x,y index locations.

In [1]:
# Make sure pandas is loaded
import pandas as pd

In [2]:
red_df = pd.read_csv('C:/Users/itspark/Documents/Analytics/dataset/red_wine.csv')
red_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [34]:
# # Read the two wine data from CSV
red_df = pd.read_csv('C:/Users/itspark/Documents/Analytics/dataset/red_wine.csv')
white_df = pd.read_csv(r'C:\Users\itspark\Documents\Analytics\dataset/white_wine.csv')
white_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [35]:
white_df.loc[:,'citric acid':'total sulfur dioxide'] #last range included
white_df.iloc[:,2:7] #last range excluded

citric acid  residual sugar  chlorides  free sulfur dioxide  \
0            0.36            20.7      0.045                 45.0   
1            0.34             1.6      0.049                 14.0   
2            0.40             6.9      0.050                 30.0   
3            0.32             8.5      0.058                 47.0   
4            0.32             8.5      0.058                 47.0   
...           ...             ...        ...                  ...   
4893         0.29             1.6      0.039                 24.0   
4894         0.36             8.0      0.047                 57.0   
4895         0.19             1.2      0.041                 30.0   
4896         0.30             1.1      0.022                 20.0   
4897         0.38             0.8      0.020                 22.0   

      total sulfur dioxide  
0                    170.0  
1                    132.0  
2                     97.0  
3                    186.0  
4                    186.0  
...                    ...  
4893                  92.0  
4894                 168.0  
4895                 111.0  
4896                 110.0  
4897                  98.0  

[4898 rows x 5 columns]

### Selecting data using Labels (Column Headings)

We use square brackets [] to select a subset of a Python object. For example, we can select all data from a column named 'quality' from the red_df DataFrame by name. There are two ways to do this:

In [36]:
# Method 1: select a 'subset' of the data using the column name
red_df['quality']

# Method 2: use the column name as an 'attribute'; gives the same output
red_df.quality

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

In [37]:
# TIP: use the .head() method we saw earlier to make output shorter
red_df['quality'].head()
#red_df.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

In [38]:
# Creates an object, red_quality, that only contains the `quality` column
red_quality = red_df['quality']
red_quality.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

In [39]:
# Check the object type for red_quality
type(red_quality)

pandas.core.series.Series

We can pass a list of column names too, as an index to select columns in that order. This is useful when we need to reorganize our data. We us two square brackets [[]]

In [40]:
# Select the quality and fixed acidity from the DataFrame
red_df[['quality','fixed acidity']].head()

quality  fixed acidity
0        5            7.4
1        5            7.8
2        5            7.8
3        6           11.2
4        5            7.4

### Extracting Range based Subsets: Slicing

Python uses 0-based indexing

### Challenge 4 - Extracting data from red_df

1. Last two index
2. First two index

In [41]:
# Start at -2 till end
red_df[-2:]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1597     10.2        5  
1598     11.0        6

In [42]:
# [start:<end>]
red_df[:2] #last range excluded

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70          0.0             1.9      0.076   
1            7.8              0.88          0.0             2.6      0.098   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   

   alcohol  quality  
0      9.4        5  
1      9.8        5

### Subsetting first row and first column

In [43]:
# dataframe[row selection][column selection] #df.loc[row,col]
red_df[0:1]['fixed acidity']

0    7.4
Name: fixed acidity, dtype: float64

# Slicing Subsets of Rows and Columns in Python

We can select specific ranges of our data in both the row and column directions using either label or integer-based indexing.

loc is primarily label based indexing. Integers may be used but they are interpreted as a label.

iloc is primarily integer based indexing


### Challenge 5 - Slicing data

What happens when you call:

red_df.iloc[0:4, 1:4]
red_df.loc[0:4, 1:4]

How to debug last code?

In [44]:
# iloc is primarily integer based indexing
red_df.iloc[0:4, 1:4] #last range excluded

volatile acidity  citric acid  residual sugar
0              0.70         0.00             1.9
1              0.88         0.00             2.6
2              0.76         0.04             2.3
3              0.28         0.56             1.9

In [45]:
red_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [46]:
red_df.loc[2,'citric acid']

0.04

In [47]:
red_df.iloc[2,2]

0.04

In [48]:
# loc is primarily label based indexing
#red_df.loc[0:4, 1:4] # Need row and column labels not the index
#red_df.loc[0:4,:]
#or
red_df.loc[0:4]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [49]:
red_df.loc[0:4, ['volatile acidity','citric acid','residual sugar']]
#red_df.loc[0:4, 'volatile acidity','citric acid','residual sugar']

volatile acidity  citric acid  residual sugar
0              0.70         0.00             1.9
1              0.88         0.00             2.6
2              0.76         0.04             2.3
3              0.28         0.56             1.9
4              0.70         0.00             1.9

Indexing by labels loc differs from indexing by integers iloc. With loc, both the start bound and the stop bound are inclusive. When using loc, integers can be used, but the integers refer to the index label and not the position. For example, using loc and select 1:4 will get a different result than using iloc to select rows 1:4.

We can also select a specific data value using a row and column location within the DataFrame and iloc indexing

In [50]:
red_df.iloc[2, 3]  

2.3

### Subsetting Data using Criteria

In [51]:
red_df[red_df['residual sugar'] == 2.3]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2               7.8             0.760         0.04             2.3      0.092   
21              7.6             0.390         0.31             2.3      0.082   
23              8.5             0.490         0.11             2.3      0.084   
32              8.3             0.655         0.12             2.3      0.083   
58              7.8             0.590         0.18             2.3      0.076   
...             ...               ...          ...             ...        ...   
1544            8.4             0.370         0.43             2.3      0.063   
1550            7.1             0.680         0.00             2.3      0.087   
1551            7.1             0.670         0.00             2.3      0.083   
1592            6.3             0.510         0.13             2.3      0.076   
1596            6.3             0.510         0.13             2.3      0.076   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2                    15.0                  54.0  0.99700  3.26       0.65   
21                   23.0                  71.0  0.99820  3.52       0.65   
23                    9.0                  67.0  0.99680  3.17       0.53   
32                   15.0                 113.0  0.99660  3.17       0.66   
58                   17.0                  54.0  0.99750  3.43       0.59   
...                   ...                   ...      ...   ...        ...   
1544                 12.0                  19.0  0.99550  3.17       0.81   
1550                 17.0                  26.0  0.99783  3.45       0.53   
1551                 18.0                  27.0  0.99768  3.44       0.54   
1592                 29.0                  40.0  0.99574  3.42       0.75   
1596                 29.0                  40.0  0.99574  3.42       0.75   

      alcohol  quality  
2         9.8        5  
21        9.7        5  
23        9.4        5  
32        9.8        5  
58       10.0        5  
...       ...      ...  
1544     11.2        7  
1550      9.5        5  
1551      9.4        5  
1592     11.0        6  
1596     11.0        6  

[109 rows x 12 columns]

In [52]:
# Using single Criteria
red_df[red_df['residual sugar'] == 2.3]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2               7.8             0.760         0.04             2.3      0.092   
21              7.6             0.390         0.31             2.3      0.082   
23              8.5             0.490         0.11             2.3      0.084   
32              8.3             0.655         0.12             2.3      0.083   
58              7.8             0.590         0.18             2.3      0.076   
...             ...               ...          ...             ...        ...   
1544            8.4             0.370         0.43             2.3      0.063   
1550            7.1             0.680         0.00             2.3      0.087   
1551            7.1             0.670         0.00             2.3      0.083   
1592            6.3             0.510         0.13             2.3      0.076   
1596            6.3             0.510         0.13             2.3      0.076   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2                    15.0                  54.0  0.99700  3.26       0.65   
21                   23.0                  71.0  0.99820  3.52       0.65   
23                    9.0                  67.0  0.99680  3.17       0.53   
32                   15.0                 113.0  0.99660  3.17       0.66   
58                   17.0                  54.0  0.99750  3.43       0.59   
...                   ...                   ...      ...   ...        ...   
1544                 12.0                  19.0  0.99550  3.17       0.81   
1550                 17.0                  26.0  0.99783  3.45       0.53   
1551                 18.0                  27.0  0.99768  3.44       0.54   
1592                 29.0                  40.0  0.99574  3.42       0.75   
1596                 29.0                  40.0  0.99574  3.42       0.75   

      alcohol  quality  
2         9.8        5  
21        9.7        5  
23        9.4        5  
32        9.8        5  
58       10.0        5  
...       ...      ...  
1544     11.2        7  
1550      9.5        5  
1551      9.4        5  
1592     11.0        6  
1596     11.0        6  

[109 rows x 12 columns]

In [53]:
# Using multiple criteria
red_df[(red_df['residual sugar'] == 2.3) & (red_df['total sulfur dioxide'] >= 100)]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
32             8.3             0.655         0.12             2.3      0.083   
95             4.7             0.600         0.17             2.3      0.058   
522            8.2             0.390         0.49             2.3      0.099   
637            9.5             0.885         0.27             2.3      0.084   
684            9.8             0.980         0.32             2.3      0.078   
772            9.5             0.570         0.27             2.3      0.082   

     free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
32                  15.0                 113.0  0.99660  3.17       0.66   
95                  17.0                 106.0  0.99320  3.85       0.60   
522                 47.0                 133.0  0.99790  3.38       0.99   
637                 31.0                 145.0  0.99780  3.24       0.53   
684                 35.0                 152.0  0.99800  3.25       0.48   
772                 23.0                 144.0  0.99782  3.27       0.55   

     alcohol  quality  
32       9.8        5  
95      12.9        6  
522      9.8        5  
637      9.4        5  
684      9.4        5  
772      9.4        5

We can use the syntax below when querying data by criteria from a DataFrame. Experiment with selecting various subsets of the “wine” data.

Equals: ==
Not equals: !=
Greater than, less than: > or <
Greater than or equal to >=
Less than or equal to <=

### Challenge 6 - Queries

1. Select a subset of rows in the red_df DataFrame that contain data alcohol > 10 and that contain fixed acidity values less than or equal to 8. How many rows did you end up with? What did your neighbour get?

2. Select a subset of rows with quality in 5 and 8. Hint: use isin function

In [54]:
red_df[(red_df['alcohol'] > 10) & (red_df['fixed acidity'] <= 8)]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
9               7.5             0.500         0.36             6.1      0.071   
11              7.5             0.500         0.36             6.1      0.071   
30              6.7             0.675         0.07             2.4      0.089   
31              6.9             0.685         0.00             2.5      0.105   
36              7.8             0.600         0.14             2.4      0.086   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
9                    17.0                 102.0  0.99780  3.35       0.80   
11                   17.0                 102.0  0.99780  3.35       0.80   
30                   17.0                  82.0  0.99580  3.35       0.54   
31                   22.0                  37.0  0.99660  3.46       0.57   
36                    3.0                  15.0  0.99750  3.42       0.60   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
9        10.5        5  
11       10.5        5  
30       10.1        5  
31       10.6        6  
36       10.8        6  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[455 rows x 12 columns]

In [55]:
a = [3,4,5,5,5,3]
5 in a

True

In [56]:
red_df[red_df['quality'].isin([5,8])]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
4               7.4             0.700         0.00             1.9      0.076   
5               7.4             0.660         0.00             1.8      0.075   
...             ...               ...          ...             ...        ...   
1582            6.1             0.715         0.10             2.6      0.053   
1583            6.2             0.460         0.29             2.1      0.074   
1589            6.6             0.725         0.20             7.8      0.073   
1594            6.2             0.600         0.08             2.0      0.090   
1597            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
4                    11.0                  34.0  0.99780  3.51       0.56   
5                    13.0                  40.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1582                 13.0                  27.0  0.99362  3.57       0.50   
1583                 32.0                  98.0  0.99578  3.33       0.62   
1589                 29.0                  79.0  0.99770  3.29       0.54   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1597                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
4         9.4        5  
5         9.4        5  
...       ...      ...  
1582     11.9        5  
1583      9.8        5  
1589      9.2        5  
1594     10.5        5  
1597     10.2        5  

[699 rows x 12 columns]

### Sort and Rank dataframes in Pandas

In [57]:
red_df.sort_values(by = 'fixed acidity')

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
45             4.6             0.520         0.15             2.1      0.054   
95             4.7             0.600         0.17             2.3      0.058   
821            4.9             0.420         0.00             2.1      0.048   
588            5.0             0.420         0.24             2.0      0.060   
94             5.0             1.020         0.04             1.4      0.045   
..             ...               ...          ...             ...        ...   
555           15.5             0.645         0.49             4.2      0.095   
554           15.5             0.645         0.49             4.2      0.095   
442           15.6             0.685         0.76             3.7      0.100   
557           15.6             0.645         0.49             4.2      0.095   
652           15.9             0.360         0.65             7.5      0.096   

     free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
45                   8.0                  65.0  0.99340  3.90       0.56   
95                  17.0                 106.0  0.99320  3.85       0.60   
821                 16.0                  42.0  0.99154  3.71       0.74   
588                 19.0                  50.0  0.99170  3.72       0.74   
94                  41.0                  85.0  0.99380  3.75       0.48   
..                   ...                   ...      ...   ...        ...   
555                 10.0                  23.0  1.00315  2.92       0.74   
554                 10.0                  23.0  1.00315  2.92       0.74   
442                  6.0                  43.0  1.00320  2.95       0.68   
557                 10.0                  23.0  1.00315  2.92       0.74   
652                 22.0                  71.0  0.99760  2.98       0.84   

     alcohol  quality  
45      13.1        4  
95      12.9        6  
821     14.0        7  
588     14.0        8  
94      10.5        4  
..       ...      ...  
555     11.1        5  
554     11.1        5  
442     11.2        7  
557     11.1        5  
652     14.9        5  

[1599 rows x 12 columns]

In [58]:
red_df.sort_values(by = 'fixed acidity',ascending=False)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
652            15.9             0.360         0.65             7.5      0.096   
557            15.6             0.645         0.49             4.2      0.095   
442            15.6             0.685         0.76             3.7      0.100   
555            15.5             0.645         0.49             4.2      0.095   
554            15.5             0.645         0.49             4.2      0.095   
...             ...               ...          ...             ...        ...   
94              5.0             1.020         0.04             1.4      0.045   
1321            5.0             0.740         0.00             1.2      0.041   
821             4.9             0.420         0.00             2.1      0.048   
95              4.7             0.600         0.17             2.3      0.058   
45              4.6             0.520         0.15             2.1      0.054   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
652                  22.0                  71.0  0.99760  2.98       0.84   
557                  10.0                  23.0  1.00315  2.92       0.74   
442                   6.0                  43.0  1.00320  2.95       0.68   
555                  10.0                  23.0  1.00315  2.92       0.74   
554                  10.0                  23.0  1.00315  2.92       0.74   
...                   ...                   ...      ...   ...        ...   
94                   41.0                  85.0  0.99380  3.75       0.48   
1321                 16.0                  46.0  0.99258  4.01       0.59   
821                  16.0                  42.0  0.99154  3.71       0.74   
95                   17.0                 106.0  0.99320  3.85       0.60   
45                    8.0                  65.0  0.99340  3.90       0.56   

      alcohol  quality  
652      14.9        5  
557      11.1        5  
442      11.2        7  
555      11.1        5  
554      11.1        5  
...       ...      ...  
94       10.5        4  
1321     12.5        6  
821      14.0        7  
95       12.9        6  
45       13.1        4  

[1599 rows x 12 columns]

### Challenge 7 - Create a new variable pH_Rank in red_df and white_df
Check for min max rank

Pandas Dataframe.rank() method returns a rank of every respective index of a series passed. The rank is returned on the basis of position after sorting.
or
rank() function compute numerical data ranks (1 through n) along axis. Equal values are assigned a rank that is the average of the ranks of those values. 

In [59]:
red_df['pH'].head()

0    3.51
1    3.20
2    3.26
3    3.16
4    3.51
Name: pH, dtype: float64

In [60]:
red_df['pH_Rank'] = red_df['pH'].rank() 
red_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank  
0      9.4        5   1439.0  
1      9.8        5    369.0  
2      9.8        5    577.0  
3      9.8        6    249.5  
4      9.4        5   1439.0

In [61]:
white_df['pH_Rank'] = white_df['pH'].rank()
white_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  pH_Rank  
0      8.8        6    474.5  
1      9.5        6   3845.0  
2     10.1        6   3493.5  
3      9.9        6   2633.0  
4      9.9        6   2633.0

In [62]:
red_df.shape

(1599, 13)

In [63]:
red_df['pH_Rank'].describe()

count    1599.000000
mean      800.000000
std       461.624311
min         1.000000
25%       406.500000
50%       803.000000
75%      1194.000000
max      1598.500000
Name: pH_Rank, dtype: float64

In [64]:
white_df.shape

(4898, 13)

In [65]:
white_df['pH_Rank'].describe()

count    4898.0000
mean     2449.5000
std      1413.7399
min         1.0000
25%      1268.5000
50%      2491.5000
75%      3673.0000
max      4898.0000
Name: pH_Rank, dtype: float64

# Combining DataFrames with Pandas



### Questions

Can I work with data from multiple sources?

How can I combine data from different data sets?

### Objectives

Combine data from multiple files into a single DataFrame using merge and concat.

Combine two DataFrames using a unique ID found in both DataFrames.

Employ to_csv to export a DataFrame in CSV format.

### Identifying common series in two dataframes

In [66]:
print(red_df.columns)
print(red_df.shape)

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'pH_Rank'],
      dtype='object')
(1599, 13)


In [67]:
print(white_df.columns)
print(white_df.shape)

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'pH_Rank'],
      dtype='object')
(4898, 13)


Create Color Columns

Create a new series with all the rows in the red and white dataframes that repeat the value “red” or “white.” 


In [68]:
red_df['color'] = 'red'
white_df['color'] = 'white'

In [69]:
red_df.shape

(1599, 14)

In [70]:
white_df.shape

(4898, 14)

In [71]:
# Check if the new series is created
red_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank color  
0      9.4        5   1439.0   red  
1      9.8        5    369.0   red  
2      9.8        5    577.0   red  
3      9.8        6    249.5   red  
4      9.4        5   1439.0   red

In [72]:
# Check if the new series is created
white_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  pH_Rank  color  
0      8.8        6    474.5  white  
1      9.5        6   3845.0  white  
2     10.1        6   3493.5  white  
3      9.9        6   2633.0  white  
4      9.9        6   2633.0  white

### Combine DataFrames with Append

Check the documentation for Pandas' append function and see if you can use this to figure out how to combine the dataframes. (Bonus: Why aren't we using the merge method to combine the dataframes?) We will come back to this later.


In [73]:
# append dataframes
wine_df = white_df.append(red_df) 

# view dataframe to check for success
wine_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  pH_Rank  color  
0      8.8        6    474.5  white  
1      9.5        6   3845.0  white  
2     10.1        6   3493.5  white  
3      9.9        6   2633.0  white  
4      9.9        6   2633.0  white

In [74]:
# view dataframe to check for success
wine_df.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  pH_Rank color  
1594     10.5        5   1335.5   red  
1595     11.2        6   1455.5   red  
1596     11.0        6   1263.0   red  
1597     10.2        5   1521.5   red  
1598     11.0        6   1151.5   red

In [80]:
4898 +1599
red_df.shape

(1599, 14)

In [77]:
wine_df.shape

(6497, 14)

In [78]:
wine_df['color'].value_counts()

white    4898
red      1599
Name: color, dtype: int64

### Concatenating DataFrames

We can use the concat function in pandas to append either columns or rows from one DataFrame to another. Let’s grab two subsets of our data to see how this works.

When we concatenate DataFrames, we need to specify the axis. axis=0 tells pandas to stack the second DataFrame under the first one. It will automatically detect whether the column names are the same and will stack accordingly. axis=1 will stack the columns in the second DataFrame to the RIGHT of the first DataFrame. To stack the data vertically, we need to make sure we have the same columns and associated column format in both datasets. When we stack horizonally, we want to make sure what we are doing makes sense (ie the data are related in some way).

In [ ]:
# 1599,14
# 4896,14
# axis=0 1599+4896
# axis=1 = 14+14

In [26]:
# Stack the DataFrames on top of each other
wine_df_concat = pd.concat([red_df,white_df], axis=0)

In [37]:
wine_df_concat.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank color  
0      9.4        5   1439.0   red  
1      9.8        5    369.0   red  
2      9.8        5    577.0   red  
3      9.8        6    249.5   red  
4      9.4        5   1439.0   red

In [63]:
wine_df_concat.shape

(6497, 14)

In [64]:
wine_df_concat['color'].value_counts()

white    4898
red      1599
Name: color, dtype: int64

### Index Reset and Drop rows or columns

In [65]:
# Check whether the earlier append went well
max(wine_df_concat.index)

4897

In [68]:
wine_df_concat[4899:]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
3300            7.8              0.25         0.37             1.0      0.043   
3301            5.6              0.12         0.33             2.9      0.044   
3302            6.6              0.24         0.28             6.7      0.032   
3303            6.3              0.22         0.34             5.0      0.032   
3304            6.0              0.32         0.30             1.9      0.033   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
3300                 10.0                  80.0  0.99128  3.08       0.38   
3301                 21.0                  73.0  0.98896  3.17       0.32   
3302                 26.0                  91.0  0.99172  3.13       0.32   
3303                 36.0                  93.0  0.99012  3.27       0.36   
3304                 41.0                 142.0  0.98912  3.29       0.42   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  pH_Rank  color  
3300     11.4        5   1154.5  white  
3301     12.9        8   2360.5  white  
3302     12.3        6   1768.0  white  
3303     13.5        7   3585.5  white  
3304     12.8        7   3757.5  white  
...       ...      ...      ...    ...  
4893     11.2        6   3585.5  white  
4894      9.6        5   2066.5  white  
4895      9.4        6    403.5  white  
4896     12.8        7   4145.0  white  
4897     11.8        6   3493.5  white  

[1598 rows x 14 columns]

The dataframe has 6497 rows- so the max index value should be 6496 (staring from 0)

In [69]:
# Lets reset the index values
wine_df.reset_index(drop=True)
wine_df.head()
# Seems it did not got stored

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank color  
0      9.4        5   1439.0   red  
1      9.8        5    369.0   red  
2      9.8        5    577.0   red  
3      9.8        6    249.5   red  
4      9.4        5   1439.0   red

In [71]:
# We need to reset the index and store the changes (overwrite the changes in original dataframe)
wine_df.reset_index(drop=True,inplace=True)
wine_df[4897:]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4897            6.3              0.24         0.35             2.3      0.039   
4898            6.8              0.32         0.32             8.7      0.029   
4899            7.8              0.25         0.37             1.0      0.043   
4900            5.6              0.12         0.33             2.9      0.044   
4901            6.6              0.24         0.28             6.7      0.032   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4897                 43.0                 109.0  0.99056  3.34       0.44   
4898                 31.0                 105.0  0.99146  3.00       0.34   
4899                 10.0                  80.0  0.99128  3.08       0.38   
4900                 21.0                  73.0  0.98896  3.17       0.32   
4901                 26.0                  91.0  0.99172  3.13       0.32   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  pH_Rank  color  
4897     11.8        6   4145.0  white  
4898     12.3        7    474.5  white  
4899     11.4        5   1154.5  white  
4900     12.9        8   2360.5  white  
4901     12.3        6   1768.0  white  
...       ...      ...      ...    ...  
6492     11.2        6   3585.5  white  
6493      9.6        5   2066.5  white  
6494      9.4        6    403.5  white  
6495     12.8        7   4145.0  white  
6496     11.8        6   3493.5  white  

[1600 rows x 14 columns]

In [72]:
# Drop row index
wine_df_drop1 = wine_df.drop(6496,axis=0) # axis = 0 for rows
wine_df_drop1.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
6491            6.5              0.23         0.38             1.3      0.032   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
6491                 29.0                 112.0  0.99298  3.29       0.54   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   

      alcohol  quality  pH_Rank  color  
6491      9.7        5   3757.5  white  
6492     11.2        6   3585.5  white  
6493      9.6        5   2066.5  white  
6494      9.4        6    403.5  white  
6495     12.8        7   4145.0  white

In [44]:
wine_df_drop2 = wine_df.drop('pH_Rank',axis=1) # axis = 1 for columns
wine_df_drop2.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  color  
6492     11.2        6  white  
6493      9.6        5  white  
6494      9.4        6  white  
6495     12.8        7  white  
6496     11.8        6  white

### Edit String variable in dataframe

### .str.upper / .str.lower

In [51]:
red_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 14 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
color                   1599 non-null object
pH_Rank                 1599 non-null float64
dtypes: float64(12), int64(1), object(1)
memory usage: 175.0+ KB


In [73]:
wine_df.color = wine_df.color.str.capitalize() #upper()
wine_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank color  
0      9.4        5   1439.0   Red  
1      9.8        5    369.0   Red  
2      9.8        5    577.0   Red  
3      9.8        6    249.5   Red  
4      9.4        5   1439.0   Red

In [46]:
wine_df.color = wine_df.color.str.lower()
wine_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  pH_Rank color  
0      9.4        5   1439.0   red  
1      9.8        5    369.0   red  
2      9.8        5    577.0   red  
3      9.8        6    249.5   red  
4      9.4        5   1439.0   red

### Writing Out Data to CSV

We can use the to_csv command to do export a DataFrame in CSV format. Note that the code below will by default save the data into the current working directory. We can save it to a different folder by adding the foldername and a slash to the file  wine_df_concat_vertical.to_csv('foldername/out.csv'). We use the ‘index=False’ so that pandas doesn’t include the index number for each line.

In [27]:
# Write DataFrame to CSV
#wine_df_concat.to_csv('wine_df_concat.csv', index=False)
wine_df_concat.to_csv('C:/Users/itspark/Documents/Analytics/dataset/winedata_concat.csv')

Check out your working directory to make sure the CSV wrote out properly, and that you can open it! If you want, try to bring it back into Python to make sure it imports properly.

In [48]:
#wine_df_concat.to_csv('G:/DS files/wine_df_concat.csv', index=False)